### ISSUE TITLE
nsis-maven-plugin

### ISSUE THREAD

**Issue by Nadahar** [UniversalMediaServer/UniversalMediaServer#724]
@UniversalMediaServer/developers @SubJunk 

While trying to clean up out POM a bit, I've stumbled upon something that might become a problem. We rely on the `nsis-maven-plugin` for our Windows build. This used to be hosted at [codehaus.org](http://codehaus.org), which terminated all its services 31. May 2015. Not only is the website down, but the SVN is down as well, so there's not even any access to the sources.

A new website has been created for some of the plugins at [mojohaus.org](http://www.mojohaus.org/plugins.html) but it is very unfinished (and has been for some months) and most of the links are dead. All links for the `nsis-maven-plugin` there are dead. They write that they are "in the process" of transferring "selected" plugins from codehaus, but it's unclear if that's still work in progress and if they have secured all the sources. Worse yet, the `nsis-maven-plugin` was never completed and we're still using the 1.0-SNAPSHOT version. Chances are that this version is gone forever. The repository used in our POM for this plugin, [Codehaus Snapshots](http://nexus.codehaus.org/snapshots/) is dead.

That said, we should never rely on SNAPSHOT dependencies as they can change without notice. So, I started to wonder how this is working at all, and by disabling and enabling repositories I found out that it's actually still hosted at [oss.sonatype.org/content/repositories/snapshots](https://oss.sonatype.org/content/repositories/snapshots). But, with the sources gone and probably noone "owning" this, it doesn't feel good to rely on.

There's another version of this plugin available av Maven central, [nsis-maven-plugin:1.0-alpha-1](http://mvnrepository.com/artifact/org.codehaus.mojo/nsis-maven-plugin/1.0-alpha-1). I tried to use this version in our POM instead, but it won't work. They have renamed their goals, but after substituing the goals with what I'm guessing is their changed names, it complains that our nsis-script has illegal characters in it. In short, it seems like a bugged version.

So, I started to search for any sources for this plugin, and I found one right here on GitHub. I've now [forked that repo](https://github.com/Nadahar/nsis-maven-plugin) to make sure it's available. The fork doesn't contain the history though, so there's no way to go back and see what's changed since the version we're using. The version of these sources are 1.0-alpha-5. This version is not available from any Maven repo, so I haven't tested it out to see if it will work with our current script.

As I see it, with the sources secured we should at least have the posibility to get it working if the last hosted copy of the 1.0-SNAPSHOT were to disappear. I could then deploy it under com.universalmediaserver to Maven central so it would be available to us and anyone else needing it.

From my previous two journeys of updating existing Maven plugins I no longer naively believes that this is quick or doesn't represent a lot of hurdles. It probably does, and there's probably a whole lot of to me yet unknown territory I have to understand to be able to do anything useful with the code. I'm therefore sceptical if I should attempt it, and if it's worth the hazzle. Nothing is certain here, it might be that it will pop up at MojoHaus in the future and all is well, or it might never reappear.

I wanted to get your thoughts about the issue. Should we just calmly wait and hope that it will stay available like it is? Should we bet on MojoHaus getting it up and running in the future? Or should we take steps ourselves to try to secure a working version of the plugin?

**Comment by Sami32** [comment]
@Nadahar So i guess it's why you didn't used _maven-external-dependency-plugin 0.6-SNAPSHOT_ ;-)

**Comment by Nadahar** [comment]
@Sami32 Not sure what you mean. Are you referring to [external-maven-plugin](https://github.com/UniversalMediaServer/external-maven-plugin) or [maven-external-dependency-plugin](https://code.google.com/archive/p/maven-external-dependency-plugin/)? None of them have any 0.6 version. Maven rules dictate that when you fork a project it should get a different name and it's own versioning, so that's what I did. Maven rules also dictate that SNAPSHOT versions can be changed at any time without a new version number whereas non SNAPSHOT versions can't be changed once distributed. Because of this we should never use SNAPSHOT versions, they can break at any time without changing version. Chances are in many instances that they won't be changed though, but the author has no obligation not to.

**Comment by Sami32** [comment]
@Nadahar 
```
<groupId>com.savage7.maven.plugins</groupId>
	<artifactId>maven-external-dependency-plugin-test</artifactId>
	<packaging>jar</packaging>
	<version>0.6-SNAPSHOT</version>
	<name>maven-external-dependency-plugin-test</name>
	<url>http://code.google.com/p/maven-external-dependency-plugin/</url>
	
```

**Comment by Nadahar** [comment]
@Sami32 I'm still not sure what you mean. I know that UMS uses SNAPSHOT dependencies, I'm just saying that it's a bad/unsafe practice. Regarding ´´´maven-external-dependency-plugin``` that particular plugin isn't used by UMS anymore as I forked and partially rewrote it.

**Comment by Sami32** [comment]
@Nadahar It's why when reading your first comment here i ask you if it was the reason why you didn't use the latest version, and the fact that you told me that 0.6 didn't exist could be the answer.

Since i saw that you make your own some month ago and didn't use the latest version, i was just enquiring about.
Why i ask you now and not before ? because one commit of @atamariya upgrade it to the latest stable 0.5 version, so i just would like know if it'll broken something or not.

**Comment by Sami32** [comment]
@Nadahar My bad 8-( 
He replaced your _external-maven-plugin_ by the latest stable _maven-external-dependency-plugin_.
I need to test it, but i think it's why i get errors with external plugins while compiling.

**Comment by Nadahar** [comment]
@atamariya Is there any particular reason why you went back to the savage7 plugin? They do pretty much the same, but ```external-maven-plugin``` is upgraded to use much fewer deprecated Maven calls (the savage7 one hasn't been updated for years and is made for a much older version of Maven). In addition to that I've simplified the flow combining two operations to one (download + install is just install) and improved error messages in many intances.

**Comment by atamariya** [comment]
@Nadahar I didn't know the history behind this change. I usually prefer the plugin in core maven repo with a higher version. 0.1 seemed a first iteration!!! However, I've reverted that back now.

**Comment by Nadahar** [comment]
@atamariya I made significant changes to it although it's a fork of the other one. Also in case the original author wanted to update the original one time in the future, I couldn't "claim" my version to be the next version. It's become "illegal" to start a plugin name with ```maven``` unless it's an official Apache one and a goal for me was to shorten the name (to make it easier to run). All in all, I reset the version number since it is to be considered a "separate artifact" though it's a fork. It is the first *release*, but by no means the first version of the plugin I was testing :wink: 

In general comparing version numbers between applications or artifacts make no sense at all. Some authors "cheat" by making big jumps in version numbering to make a product seem more mature or something, while others have a strictly functional version system. It's very much like comparing apples and oranges.

**Comment by vicnevicne** [comment]
Hi and sorry to dig out this old bug.
I'm trying to build UMS on Windows and stumbled upon the "[ERROR] Plugin org.codehaus.mojo:nsis-maven-plugin:1.0-SNAPSHOT or one of its dependencies could not be resolved: Could not find artifact org.codehaus.mojo:nsis-maven-plugin:jar:1.0-SNAPSHOT" message.
I am a maven newbie, but what is the recommended path to solve the issue ? Should I build the nsis.jar by hand from the project you forked ? Or can anyone share the SNAPSHOT binary that was once used ?
I find it sad that it seems impossible to build UMS from scratch :-(
Any hint is highly welcome.

**Comment by Nadahar** [comment]
@vicnevicne Are you using latest master when trying to build? It should be possible to build UMS from scratch, we do it all the time using [Travis-CI](https://travis-ci.org/). Some times some artifacts is unavailable, but it usually doesn't last very long.

**Comment by vicnevicne** [comment]
Wow, that was a prompt reply :-). Thanks.
I followed the instructions at https://github.com/UniversalMediaServer/UniversalMediaServer/wiki/Build so basically:
```
git clone git://github.com/universalmediaserver/universalmediaserver.git
cd universalmediaserver
mvn external:install
mvn package
```
Is that the right procedure ?

**Comment by Nadahar** [comment]
@vicnevicne It seems correct, but as I said, some times some of the artifacts aren't available/the server that host them doesn't reply or is offline. You only need to repeat the last step, ```mvn package``` to build again - so just try again and see if it succeeds. [This very recent build](https://travis-ci.org/UniversalMediaServer/UniversalMediaServer/jobs/213521280) indicates that all artifacts should be available.

**Comment by vicnevicne** [comment]
Hmm, the build log you refer to doesn't seem to show any reference to the "nsis-maven-plugin:jar" snapshot that causes my error (hence the reason why I resurrected this issue). 
And indeed mojo.codehaus.org seems to be dead for long. Or is it back most of the time but just down today ?

**Comment by Nadahar** [comment]
@vicnevicne Sorry, I wasn't thinking straight. "nsis-maven-plugin" is only needed when building for Windows, which Travis-CI doesn't do. It might have been unavailable for a long time with none of us noticing, because we already have the artifact cached locally.

[Maven-NSIS-2.1.zip](https://github.com/UniversalMediaServer/UniversalMediaServer/files/859813/Maven-NSIS-2.1.zip)

I've uploaded the artifact here for now. You need to extract this into your local Maven repository (```.M2```) so that the path becomes ```.m2\repository\maven\maven-nsis-plugin\2.1``` where the files are places inside the ```2.1``` folder. That should trick Maven into thinking that it's already installed.

We definitly need to do something about this now, as it seems my predictions were right.

**Comment by vicnevicne** [comment]
Thanks for your time and thanks for the zip.
Unfortunately, that didn't solve the issue yet. 
I first tried putting the contents of the zip inside .m2\repository\**maven**\maven-nsis-plugin\2.1 as you said with no success.
I then thought it was strange that I had to create the "maven" folder in bold above, so wondered if there wasn't a typo and the folder shouldn't be .m2\repository\**org\apache\maven**\maven-nsis-plugin\2.1, to no avail so I moved it back.
But are we talking about the right file ? My build complains about artifactId "nsis-maven-plugin" with groupId org.codehaus.mojo while you have provided "maven-nsis-plugin" (note the difference) with groupId "maven". 
Maybe both are missing...

**Comment by Nadahar** [comment]
@vicnevicne What a mess. It seems we have two different versions as dependencies, and I zipped the wrong version for you. Here is the missing one:

[nsis-maven-plugin.zip](https://github.com/UniversalMediaServer/UniversalMediaServer/files/859852/nsis-maven-plugin.zip)

Its path should look like this: ```.m2\repository\org\codehaus\mojo\nsis-maven-plugin\1.0-SNAPSHOT```

**Comment by Nadahar** [comment]
@vicnevicne As I discovered in the mean time I gave you the wrong artifact. The paths are correct though, I haven't typed them - they are copy-paste.

**Comment by vicnevicne** [comment]
Yes ! It is building now as I write...  Thanks a lot for your help.
If you need tests on a Windows build system, don't hesistate to ask. I'd be glad to contribute in return if I can. 
Kind regards

**Comment by Nadahar** [comment]
Fixed in #1230.
